rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2353 @@
+
+impl UnsafeTransmuteOptions for NeglectValidity {}
+```
+
+By using the `NeglectValidity` option, you are committing to ensure dynamically source value is a valid instance of the destination type. For instance:
+```rust
+#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]
+#[repr(u8)]
+enum Bool {
+    True = 1,
+    False = 0,
+}
+
+pub trait TryIntoBool
+{
+    fn try_into_bool(self) -> Option<Bool>;
+}
+
+impl<T> TryIntoBool for T
+where
+    T: TransmuteInto<u8>,
+    u8: TransmuteInto<Bool, NeglectValidity>
+{
+    fn try_into_bool(self) -> Option<Bool> {
+        let val: u8 = self.transmute_into();
+
+        if val > 1 {
+            None
+        } else {
+            // Safe, because we've first verified that
+            // `val` is a bit-valid instance of a boolean.
+            Some(unsafe {val.unsafe_transmute_into()})
+        }
+    }
+}
+```
+
+Even with `NeglectValidity`, the compiler will statically reject transmutations that cannot possibly be valid:
+```rust
+#[derive(PromiseTransmutableInto)]
+#[repr(C)] enum Foo { A = 24 }
+
+#[derive(PromiseTransmutableFrom)]
+#[repr(C)] enum Bar { Z = 42 }
+
+let _ = <Bar as TransmuteFrom<Foo, NeglectValidity>::unsafe_transmute_from(Foo::N) // Compile error!
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+
+## Implementation Guidance
+The *only* item defined by this RFC requiring special compiler support is `TransmuteFrom`. To realize this RFC's proposal of safe transmutation between different types, this item will require compiler support. However, the *most* minimal acceptable implementation of `TransmuteFrom` can be achieved entirely in-language:
+
+```rust
+/// A type is transmutable into itself.
+unsafe impl<T, Neglect> TransmuteFrom<T, Neglect> for T
+where
+    Neglect: UnsafeTransmuteOptions
+{}
+
+/// A transmutation is *stable* if...
+unsafe impl<Src, Dst> TransmuteFrom<Src> for Dst
+where
+    Src: PromiseTransmutableInto,
+    Dst: PromiseTransmutableFrom,
+    <Dst as PromiseTransmutableFrom>::Archetype:
+        TransmuteFrom<
+            <Src as PromiseTransmutableInto>::Archetype,
+            NeglectStability
+        >
+{}
+```
+This [minimal implementation][minimal-impl] is sufficient for convincing the compiler to accept basic stability declarations, such as those of Rust's primitive types. It is *insufficient* for making the compiler accept transmutations between *different* types (and, consequently, complex stability declarations). Implementers should use this as a starting point.
+
+### Listing for Initial, Minimal Implementation
+[minimal-impl]: #Listing-for-Initial-Minimal-Implementation
+**[This module listing provides the minimal-viable implementation of this RFC (excepting the automatic derives).](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=25a83d8d8e33e9f631a64b53e3c39844)** This listing is also the **canonical specification** of this RFC's public API surface.
```

Comment:
```suggestion
**[This module listing provides the minimal-viable implementation of this RFC (excepting the automatic derives).](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=bca1ad92882760e94cd4a79bccf4de30)** This listing is also the **canonical specification** of this RFC's public API surface.
```

[Simplify](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9ffa28df10593c249795ad93a7a64ebf) the minimal implementation by removing the `private` module.

This (commented-out) bound remains problematic:
```rust
    /// `Self: TransmuteInto<Dst, Neglect`, if the compiler accepts the stability,
    /// safety, and soundness of transmuting `Self` into `Dst`, notwithstanding
    /// a given set of static checks to `Neglect`.
    pub unsafe trait TransmuteInto<Dst: ?Sized, Neglect = ()>
    where
        /*Dst: TransmuteFrom<Self, Neglect>,*/ // FIXME: Why is adding this bound so problematic?
        Neglect: UnsafeTransmuteOptions,
    {
```

Reply:
I've [managed](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9cd07f2ac2ecb95f558f103435554c63) to un-comment the bound, but doing so requires a lot of [new](https://gist.github.com/jswrenn/19aad724cb2e804fae6abd31a12988ff/revisions#diff-1fb9bba94efc1d84c71b0bd852912496), seemingly-redundant bounds anywhere `TransmuteInto` appears.

Reply:
[`Into`](https://doc.rust-lang.org/std/convert/trait.Into.html) doesn't have this reflexivity bound, so maybe it's not even a desirable thing to have?